stack/corepersistence/queryindex/src/main/java/org/apache/usergrid/persistence/index/impl/IndexBatchBufferImpl.java

```diff
@@ -0,0 +1,226 @@
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.action.support.replication.ShardReplicationOperationRequestBuilder;
+import org.elasticsearch.client.Client;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+import rx.Subscriber;
+import rx.functions.Action0;
+import rx.functions.Action1;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.*;
+
+
+/**
+ * Buffer index requests into sets to send.
+ */
+@Singleton
+public class IndexBatchBufferImpl implements IndexBatchBuffer {
+
+    private static final Logger log = LoggerFactory.getLogger(IndexBatchBufferImpl.class);
+    private final MetricsFactory metricsFactory;
+    private final Counter indexSizeCounter;
+    private final Client client;
+    private final FailureMonitorImpl failureMonitor;
+    private final IndexFig config;
+    private final Timer flushTimer;
+    private final ArrayBlockingQueue<RequestBuilderContainer> blockingQueue;
+    private Observable<List<RequestBuilderContainer>> consumer;
+    private Producer producer;
+
+    @Inject
+    public IndexBatchBufferImpl(final IndexFig config, final EsProvider provider, final MetricsFactory metricsFactory){
+        this.metricsFactory = metricsFactory;
+        this.flushTimer = metricsFactory.getTimer(IndexBatchBuffer.class, "index.buffer.flush");
+        this.indexSizeCounter =  metricsFactory.getCounter(IndexBatchBuffer.class, "index.buffer.size");
+        this.config = config;
+        this.blockingQueue = new ArrayBlockingQueue<>(config.getIndexBatchSize());
+        this.failureMonitor = new FailureMonitorImpl(config,provider);
+        this.producer = new Producer();
+        this.client = provider.getClient();
+
+        consumer();
+    }
+
+    private void consumer() {
+        //batch up sets of some size and send them in batch
+        this.consumer = Observable.create(producer)
+                .buffer(config.getIndexBufferTimeout(), TimeUnit.MILLISECONDS, config.getIndexBufferSize())
+                .doOnNext(new Action1<List<RequestBuilderContainer>>() {
+                    @Override
+                    public void call(List<RequestBuilderContainer> containerList) {
+                        for (RequestBuilderContainer container : containerList) {
+                            blockingQueue.add(container);
+                        }
+                        flushTimer.time();
+                        indexSizeCounter.dec(containerList.size());
+                        execute(config.isForcedRefresh());
+                    }
+                });
+        consumer.subscribe();
+    }
+
+    @Override
+    public BetterFuture<ShardReplicationOperationRequestBuilder> put(IndexRequestBuilder builder){
+        RequestBuilderContainer container = new RequestBuilderContainer(builder);
+        metricsFactory.getCounter(IndexBatchBuffer.class,"index.buffer.size").inc();
```

Comment:
we can, i did this because i knew all the keys were cached in a singleton hashmap